### PR TITLE
GLSP-974: Use interface-injection for all subclasses of ModelState

### PR DESCRIPTION
**Pull request by CamilleLetavernier**
- Use interfaces for EMFModelState and EMFNotationModelState
- Use these interfaces for binding and injection

refs https://github.com/eclipse-glsp/glsp/issues/974

### CODE REVIEW

**Comment by planger** [comment]
Nice! Maybe is https://github.com/eclipse-glsp/glsp/issues/896 or is/could be addressed in this go?

**Comment by CamilleLetavernier** [comment]
> Maybe is https://github.com/eclipse-glsp/glsp/issues/896 or is/could be addressed in this go?

I'll take a look :+1:

**Comment by CamilleLetavernier** [comment]
I believe this is fixed by this PR, as we now bind the class only once. Extra bindings only specify additional interface-bindings for the same class. However I'm not really sure how to test it. Is there an example where I could try it?

**Comment by planger** [comment]
> I believe this is fixed by this PR, as we now bind the class only once. Extra bindings only specify additional interface-bindings for the same class.

That's what I also was thinking looking at the code.

> However I'm not really sure how to test it. Is there an example where I could try it?

Well, I think it should be sufficient if you'd remove the override of bindGModelState() in your subclass of EMFDiagramModule. The use case was basically having a concrete DiagramModule that extends EMFDiagramModule but uses the EMFModelState instead of binding a specialization of EMFModelState (in its bindGModelState() override).

File: plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/EMFModelStateImpl.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
+/********************************************************************************
+ * Copyright (c) 2022 EclipseSource and others.
```
Comment (ndoschek) [inline plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/EMFModelStateImpl.java:2]:
```suggestion
 * Copyright (c) 2022-2023 EclipseSource and others.
```

File: plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/notation/EMFNotationModelStateImpl.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
+/********************************************************************************
+ * Copyright (c) 2022 EclipseSource and others.
```
Comment (ndoschek) [inline plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/notation/EMFNotationModelStateImpl.java:2]:
```suggestion
 * Copyright (c) 2022-2023 EclipseSource and others.
```

File: plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/model/DefaultGModelState.java [hunk 3]
```diff
@@ -42,6 +44,10 @@ public class DefaultGModelState implements GModelState {
    protected String editMode;
    protected GModelIndex index = GModelIndex.empty();
 
+   public DefaultGModelState() {
+      System.err.println("Instantiate Model State " + getClass().getName());
```
Comment (ndoschek) [inline plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/model/DefaultGModelState.java:48]: I assume this line is a leftover?

File: plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/model/DefaultGModelState.java
Comment (ndoschek) [inline plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/model/DefaultGModelState.java:1]: Update year range in header.

**Review by ndoschek — COMMENTED** [review]
Thanks a lot @CamilleLetavernier!
Changes look great to me, I did not really test it, but it will be easier to test it with the already opened [PR](https://github.com/eclipse-emfcloud/modelserver-glsp-integration/pull/53#event-8982714256) in the modelserver-glsp-integration.

Besides that I only have a few very minor comments, mostly regarding the year ranges in the file headers.

File: plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/EMFModelState.java [hunk 4]
```diff
@@ -1,110 +1,26 @@
 /********************************************************************************
- * Copyright (c) 2022 EclipseSource and others.
+ * Copyright (c) 2023 EclipseSource and others.
```
[the remark this answers is not in the document]
Comment (CamilleLetavernier) [reply plugins/org.eclipse.glsp.server.emf/src/org/eclipse/glsp/server/emf/EMFModelState.java:2]: In this case it's a new interface (even though it replaces a class with the same file name), so I think it needs new copyright headers

**Review by ndoschek — APPROVED** [review]
Thanks @CamilleLetavernier for the updates! LGTM :+1:

**Comment by CamilleLetavernier** [comment]
Commit a0350bf updates the Changelog for 2.x, as this PR includes breaking changes related to EMFModelState and EMFNotationModelState

**Review by ndoschek — APPROVED** [review]

**Pull request merged**